mitodl/unified-ecommerce
# Add tax collection and blocked countries processing

### What are the relevant tickets?

Closes mitodl/hq#5766

### Description (What does it do?)

Adds a `basket_add` hook that is fired before a user adds an item to their basket, so we can perform tasks based on the user and product being added.

Adds a wrapper plugin to pull in some data that other plugins will need and finish setting up the basket object. (At the moment, this means geolocating the user for tax lookup and blocked countries reasons.) Geolocation works based on geolocating via IP and by considering the user's profile.

Adds a check for blocked countries that runs when an item is added to the basket. If the user belongs to a country that we've blocked either globally in UE or for a particular product, the check raises an exception and basket processing stops.

Adds a check for tax assessment for the user. If the user is in a country that assesses tax, we then flag the basket as being taxable with the rate that matches for the user.

Pulls in the geoip code from ol-django for geolocation, adds support for field mappings in the APISIX middleware, and adds a few other necessary things for this process. It also adds the necessary data to orders so tax is collected through the order process, and updates the display of the orders in Django Admin to make them more usable.

### Screenshots

Completed order in Django Admin (displays the new fields/groupings):

![image](https://github.com/user-attachments/assets/27474f50-6422-4813-bbce-6cc6265267a7)

Cart with some updates to show tax/etc.:

![image](https://github.com/user-attachments/assets/5173e0d1-2dbf-4c7c-ad8b-bce88083bb7a)


### How can this be tested?

You will need to import the MaxMind GeoLite2 databases. See the documentation in the ol-django geoip app for instructions.

For best results, simply add a "country code" to your user profile manually through the Django Admin.

1. Add a blocked country and a tax rate using the admin interface. Blocked countries can be for a particular product or for the entire system, so you should configure both.
3. Configure test accounts (or equivalent): one that should _not_ hit any blocks/checks, one that _should_ hit the blocked countries check globally, and one that _should_ hit the blocked countries check for a particular product. Also, configure one that will not hit the block but will be assessed tax.
4. Attempt to check out in each scenario. Your results should match the truth table below. Note that the cart test mule won't handle the error gracefully so you will likely need to check in devtools - you should get HTTP 451 Unavailable for Legal Reasons for blocked countries. 
5. For the scenario that involves tax, you should additionally see tax assessed in the cart, and you should be charged tax in the order. Tax should be applied across the entire order.

For completed orders that involve tax:
- You should see the tax appear in the CyberSource payment screens.
- The order as displayed in the Django Admin should display the calculated tax amounts.
- Discounts should be applied before tax - i.e. we collect tax on the _discounted_ amount.
- The line should _also_ display the correct amount of tax.
- The transaction should _also_ display the correct totals.

For completed orders that involve discounts:
- You should see new fields in the Django Admin for the order that break out the total amount of discount that was applied.
